### PR TITLE
Make the library fork safe

### PR DESCRIPTION
**Pull request by casperisfine**
We record the pid when establishing the connection, and compare the recorded pid with the current one whenever we re-use the connection.

If we detect a change, we clone the connection and close the old one.

We make sure not to send `QUIT` when closing the old connection as it would break the parent's connection.

NB: `libmemcached` doesn't expose a way to close a connection without sending `QUIT`, so I had to patch it...

### CODE REVIEW

**Comment by casperisfine** [comment]
This seem to be causing SEGV in on our app test suite:

```
/usr/local/ruby/bin/ruby(sigsegv+0x4d) [0x5570cd5655fd] /tmp/ruby-build.20230719070129.391.cSc12N/ruby-3.2.0-36486ba8544a5aed5ff97b24c082ceada532ccdb/signal.c:964
/lib/x86_64-linux-gnu/libc.so.6(0x7f54d4ed9520) [0x7f54d4ed9520]
/tmp/bundle/ruby/3.2.0/bundler/gems/memcached-2aac5bc7d3d0/lib/rlibmemcached.so(memcached_send+0x44) [0x7f54c3a84450] /tmp/bundle/ruby/3.2.0/bundler/gems/memcached-2aac5bc7d3d0/vendor/libmemcached-0.32/libmemcached/memcached_storage.c:78
/tmp/bundle/ruby/3.2.0/bundler/gems/memcached-2aac5bc7d3d0/lib/rlibmemcached.so(memcached_set+0x23) [0x7f54c3a849e6] /tmp/bundle/ruby/3.2.0/bundler/gems/memcached-2aac5bc7d3d0/vendor/libmemcached-0.32/libmemcached/memcached_storage.c:187
/tmp/bundle/ruby/3.2.0/bundler/gems/memcached-2aac5bc7d3d0/lib/rlibmemcached.so(SWIG_From_int+0x0) [0x7f54c3a79a89] /tmp/bundle/ruby/3.2.0/bundler/gems/memcached-2aac5bc7d3d0/ext/rlibmemcached/rlibmemcached_wrap.c:9387
/tmp/bundle/ruby/3.2.0/bundler/gems/memcached-2aac5bc7d3d0/lib/rlibmemcached.so(_wrap_memcached_set) /tmp/bundle/ruby/3.2.0/bundler/gems/memcached-2aac5bc7d3d0/ext/rlibmemcached/rlibmemcached_wrap.c:9388
/usr/local/ruby/bin/ruby(vm_call_cfunc_with_frame+0x137) [0x5570cd5f7d97] /tmp/ruby-build.20230719070129.391.cSc12N/ruby-3.2.0-36486ba8544a5aed5ff97b24c082ceada532ccdb/vm_insnhelper.c:3270
/usr/local/ruby/bin/ruby(vm_sendish+0x9c) [0x5570cd60b884] /tmp/ruby-build.20230719070129.391.cSc12N/ruby-3.2.0-36486ba8544a5aed5ff97b24c082ceada532ccdb/vm_insnhelper.c:5082
```

It's not fully clear to me what is happening yet.

**Comment by casperisfine** [comment]
So right now if you use an existing memcached connection after fork, we properly discard the parent FD and avoid sending `QUIT`.

However if you don't use that connection in the child and then exit, since `memcached_free` is defined as destructor, then `QUIT` will be emitted. 

This means that a simple:

```ruby
memcached.get("foo")
pid = fork do
  exit 0
end
Process.wait(pid)
memcached.get("foo") # QUIT was sent, connection is likely broken.
```

So I'm tempted to patch libmemcached further and just not send `QUIT` on `memcached_free`.

File: vendor/libmemcached-0.32/libmemcached/memcached.c [hunk 1]
```diff
@@ -39,8 +39,12 @@ memcached_st *memcached_create(memcached_st *ptr)
...
 void memcached_free(memcached_st *ptr)
 {
-  /* If we have anything open, lets close it now */
-  memcached_quit(ptr);
+  // memcached_quit(ptr);
+  // PATCH: The original code calls `memcached_quit`, which sends a QUIT command
+  // before closing the socket.
+  // This is problematic in forking environment as it may break inherited connections.
+  // Ideally we'd call `memcached_quit` if the connection was established in that same process.
+  memcached_discard(ptr);
```
Comment (casperisfine) [inline vendor/libmemcached-0.32/libmemcached/memcached.c]: If we kept track of the PID, we could conditionally `close` or `discard`.

**Comment by casperisfine** [comment]
Ok, so in the end I implemented the PID tracking to automatically either call the normal `QUIT + shutdown + close` procedure or to discard the FD with `dup2 + close` if the FD was inherited.

This should ensure a child process never cause the parent connection to be shutdown.

**Review by grcooper — APPROVED** [review]
The approach makes sense to me.

We should probably open a ticket to figure out why `bundle exec rake build install` isn't working.

**Pull request merged**